modules/aws-http4s/test/src-jvm/smithy4s/aws/internals/AwsSignatureTest.scala

```diff
@@ -89,12 +89,77 @@ object AwsSignatureTest extends SimpleIOSuite with Checkers {
       awsRequest: SdkHttpFullRequest
   )
   object TestInput {
-    implicit val show: Show[TestInput] = Show.fromToString
+    implicit val showContentStreamProvider: Show[ContentStreamProvider] =
+      Show.show { csp =>
+        scala.io.Source.fromInputStream(csp.newStream()).mkString
+      }
+
+    implicit val showSdkHttpFullRequest: Show[SdkHttpFullRequest] = Show.show {
+      (req: SdkHttpFullRequest) =>
+        val headers = Option(req.headers().asScala.toList)
+          .filter(_.nonEmpty)
+          .map(_.map { case (k, v) => s"$k: $v" })
+          .map(
+            _.mkString("\n        ", System.lineSeparator() + "        ", "")
+          )
+          .getOrElse(" none")
+
+        s"""(SdkHttpFullRequest)
+           |    - protocol: ${req.protocol()}
+           |    - host: ${req.host()}
+           |    - port: ${req.port()}
+           |    - method: ${req.method()}
+           |    - encodedPath: ${req.encodedPath()}
+           |    - headers:$headers
+           |    - contentStreamProvider: ${req.contentStreamProvider.toScala.show}
+           |""".stripMargin
+    }
+
+    implicit val showTestInput: Show[TestInput] = Show.show { testInput =>
+      import testInput._
+      s"""TestInput:
+         |  - serviceName: $serviceName
+         |  - operationName: $operationName
+         |  - smithy4sTimestamp: $smithy4sTimestamp
+         |  - smithy4sCredentials: $smithy4sCredentials
+         |  - smithy4sRegion: $smithy4sRegion
+         |  - awsRequest: ${awsRequest.show}
+         |""".stripMargin
+    }
 
-    val genAwsRequest = for {
+    val genPathSegment: Gen[String] =
+      Gen
+        .frequency(
+          1 -> Gen.oneOf("..", "."),
+          9 -> Gen
+            .stringOf(
+              Gen.frequency(
+                9 -> Gen.oneOf(
+                  Gen.alphaNumChar,
+                  // this list from https://github.com/http4s/http4s/blob/ac5c5cc40f732df238ef38c8ef8f571b4e66bfe3/core/shared/src/main/scala/org/http4s/internal/UriCoding.scala#L26
+                  Gen.oneOf("-_.~".toIndexedSeq)
+                ),
+                1 -> Gen.oneOf(
+                  // this list from https://github.com/http4s/http4s/blob/ac5c5cc40f732df238ef38c8ef8f571b4e66bfe3/core/shared/src/main/scala/org/http4s/Uri.scala#L1069
+                  "!$&'()*+,;=:/?@".toIndexedSeq
+                )
+              )
+            )
+        )
+        .map(
+          Uri.encode(
+            _,
+            toSkip = org.http4s.internal.CharPredicate.AlphaNum ++ "-_.~"
+          )
+        )
+
+    val genAwsRequest: Gen[SdkHttpFullRequest] = for {
       httpMethod <- Gen.oneOf(SdkHttpMethod.values().toList)
       host <- Gen.identifier
-      path <- Gen.listOfN(3, Gen.identifier).map(_.mkString("/"))
+      path <- Gen
```

Comment:
When I was working on this and trying to address some edge cases, I commented out this generator and used a variety of generators specific to the edge case I was working on. For example, things like `Gen.const("foo/../.")`, `Gen.const("this:might:be:an:aws:arn")`, etc. Unfortunately I didn't save the specifics, but they should be encoded in this generator.